macros/utils/cross_db_utils/get_profile_creation_query.sql

```diff
@@ -4,50 +4,80 @@
+GRANT SELECT ON ALL TABLES IN SCHEMA IDENTIFIER($elementary_schema_fqn) TO ROLE IDENTIFIER($elementary_role);
+GRANT SELECT ON FUTURE TABLES IN SCHEMA IDENTIFIER($elementary_schema_fqn) TO ROLE IDENTIFIER($elementary_role);
+GRANT SELECT ON ALL VIEWS IN SCHEMA IDENTIFIER($elementary_schema_fqn) TO ROLE IDENTIFIER($elementary_role);
+GRANT SELECT ON FUTURE VIEWS IN SCHEMA IDENTIFIER($elementary_schema_fqn) TO ROLE IDENTIFIER($elementary_role);
+
+-- Information schema access
+CREATE OR REPLACE PROCEDURE ELEMENTARY_GRANT_INFO_SCHEMA_ACCESS(database_name STRING, role_name STRING)
+  RETURNS VARCHAR
+  LANGUAGE SQL
+  AS
+  $$
+    BEGIN
+      GRANT USAGE,MONITOR ON DATABASE IDENTIFIER(:database_name) TO ROLE IDENTIFIER(:role_name);
+      GRANT USAGE,MONITOR ON ALL SCHEMAS IN DATABASE {{ database }} TO ROLE IDENTIFIER(:role_name);
+      GRANT USAGE,MONITOR ON FUTURE SCHEMAS IN DATABASE {{ database }} TO ROLE identifier(:role_name);
+
+      GRANT REFERENCES ON ALL TABLES IN DATABASE IDENTIFIER(:database_name) TO ROLE IDENTIFIER(:role_name);
+      GRANT REFERENCES ON ALL VIEWS IN DATABASE IDENTIFIER(:database_name) TO ROLE IDENTIFIER(:role_name);
+      GRANT REFERENCES ON ALL EXTERNAL TABLES IN DATABASE IDENTIFIER(:database_name) TO ROLE IDENTIFIER(:role_name);
+
+      GRANT REFERENCES ON FUTURE TABLES IN DATABASE IDENTIFIER(:database_name) TO ROLE IDENTIFIER(:role_name);
+      GRANT REFERENCES ON FUTURE VIEWS IN DATABASE IDENTIFIER(:database_name) TO ROLE IDENTIFIER(:role_name);
+      GRANT REFERENCES ON FUTURE EXTERNAL TABLES IN DATABASE IDENTIFIER(:database_name) TO ROLE IDENTIFIER(:role_name);
+    END;
+  $$
+;
+
+{%- set databases = elementary.get_configured_databases_from_graph() %}
+{% for database in databases -%}
+CALL ELEMENTARY_GRANT_INFO_SCHEMA_ACCESS('{{ database }}', $elementary_role);
 {% endfor %}
--- Query history views access
-GRANT DATABASE ROLE SNOWFLAKE.USAGE_VIEWER TO ROLE {{ parameters["role"] }};
-GRANT DATABASE ROLE SNOWFLAKE.GOVERNANCE_VIEWER TO ROLE {{ parameters["role"] }};
-
--- Query history access per warehouse (so Elementary can query history for queries ran by different warehouses)
-USE DATABASE {{ database }};
-CREATE OR REPLACE PROCEDURE GRANT_MONITOR_ON_ALL_WAREHOUSES(_ROLE VARCHAR) RETURNS VARCHAR
-    LANGUAGE javascript
-EXECUTE AS CALLER
-AS
-$$
-var all_warehouses = snowflake.createStatement({sqlText: `SHOW WAREHOUSES`}).execute();
-while(all_warehouses.next()) {
-  cur_warehouse = all_warehouses.getColumnValue("name");
-  snowflake.createStatement({ sqlText:`GRANT MONITOR ON WAREHOUSE ${cur_warehouse} TO ROLE ${_ROLE}`}).execute();
-}
-$$;
-CALL GRANT_MONITOR_ON_ALL_WAREHOUSES('{{ parameters["role"] }}');
+
+-- Query history access
+CREATE OR REPLACE PROCEDURE ELEMENTARY_GRANT_QUERY_HISTORY_ACCESS(role_name STRING)
+  RETURNS VARCHAR
+  LANGUAGE SQL
+  AS
+  $$
+    BEGIN
+      GRANT DATABASE ROLE SNOWFLAKE.OBJECT_VIEWER TO ROLE IDENTIFIER(:role_name);
```

Comment:
OBJECT_VIEWER technically belongs in the info schema procedure, but since it only needs to run once an not per DB it was a bit more convenient to put here :)